Added case in derive_by_array for sparse array

kangzhiq opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
#16941

#### Brief description of what is fixed or changed
The sparse array is cast to dense array while performing a derivation. Using directly sparse array will bring an improvement when it comes to a great dimension array whose value are mostly zeros. 
The sparse was cast to dense array for derivative by both **array** and **scalar**. 
Before:
```python
>>> a = MutableSparseNDimArray.zeros(10000, 20000)
>>> a[1, 1] = i
>>> a[1, 2] = j
>>> d = derive_by_array(a, i)
MemoryError
>>> d = derive_by_array(a, (i, j))
MemoryError
```
Now：
```python
>>> a = MutableSparseNDimArray.zeros(10000, 20000)
>>> a[1, 1] = i
>>> derive_by_array(a, i) = ImmutableSparseNDimArray({20001:1},(10000, 20000))
True
>>> a[1, 2] = j
>>> derive_by_array(a, (i, j)) == ImmutableSparseNDimArray({20001: 1, 200020002: 1},(2, 10000, 20000))
True
```

#### Other comments 

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
  * Added specific case for sparse array in derive_by_array 
<!-- END RELEASE NOTES -->

Commit:
Added case in derive_by_array for sparse array

- derive_by_array for sparse array will not be case to dense array

czgdp1807 commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -129,10 +129,18 @@ def f(pointer):
         if shape is None and iterable is None:
             shape = ()
             iterable = ()
+        # Construction of a sparse array from a disctionary
```
disctionary -> dictionary

czgdp1807 commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -129,10 +129,18 @@ def f(pointer):
             shape = ()
             iterable = ()
+        # Construction of a sparse array from a disctionary
+        elif shape is not None and isinstance(iterable, (dict, Dict)):
```
elif -> if as there is a return under it

kangzhiq replied:
Is it a convention in SymPy?

kangzhiq replied:
Is this a convention in SymPy?

czgdp1807 replied:
It's not about convention, it's just about avoiding the things which aren't needed. I don't know if you follow some other principles.

kangzhiq replied:
I changed the assignment of values as other blocs, so problem solved. 
But just out of curiousity, would you please explain to me why should elif -> if? I don't really see the point. As far as I know, a reasonable if-elif-else bloc ensures the logical consistency so that the code is more readable.

czgdp1807 replied:
No issues if you follow other principles. Everyone has their own coding style. It appeared non-intuitive to me to have an if just after an elif.

czgdp1807 commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -129,10 +129,18 @@ def f(pointer):
+        elif shape is not None and isinstance(iterable, (dict, Dict)):
+            return shape, iterable
+        # Construction of a sparse array from a sparse array
+        elif shape is None and isinstance(iterable, SparseNDimArray):
```
> elif -> if as there is a return under it

+1

czgdp1807 commented on sympy/tensor/array/sparse_ndim_array.py:
```diff
@@ -109,7 +109,24 @@ def reshape(self, *newshape):
+                    break
+                s_a = s_a2
+            return type(self)(s_a, self._shape)
+        else:
```
else not needed.

czgdp1807 commented on sympy/tensor/array/sparse_ndim_array.py:
```diff
@@ -109,7 +109,24 @@ def reshape(self, *newshape):
+                s_a = s_a2
+            return type(self)(s_a, self._shape)
+        else:
+            return None
```
Infact, AFAIK, any function/method in python returns `None` by default, so I think it's possible to remove, L127 and L128.

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -197,7 +198,8 @@ def derive_by_array(expr, dx):
                 raise ValueError("cannot derive by this array")
 
     if isinstance(expr, array_types):
-        expr = ImmutableDenseNDimArray(expr)
```
have you tried `.as_immutable()`?

Upabjojr commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -129,10 +129,18 @@ def f(pointer):
+        from sympy.tensor.array import SparseNDimArray
+        from sympy import Dict
 
         if shape is None and iterable is None:
```
maybe group `shape is None` into an external `if`

Upabjojr commented on sympy/tensor/array/sparse_ndim_array.py:
```diff
@@ -109,7 +109,24 @@ def reshape(self, *newshape):
-        return type(self)(*(newshape + (self._array,)))
+        return type(self)(*(newshape + (self._sparse_array,)))
+    
+    def _eval_derivative_n_times(self, s, n):
```
maybe you should not overwrite this specific method. Try to look for `_visit_eval_derivative_<...>` methods.

kangzhiq replied:
I believe that is it `al_derivative_array` that should be overridden.

Commit:
Ameliorated some codes

- regrouped if blocs in _handle_ndarray_creation_inputs
- ameliorated codes modified or added

Commit:
Modified sparse array and added __eq__

- Removed overriden function _eval_derivative_n_times
- Overrode _eval_derivative_array
- Overrode __eq__ to accelerate the equality evaluation
- Overrode __hash__ to make sparse array hashable

kangzhiq commented on sympy/tensor/array/sparse_ndim_array.py:
```diff
@@ -111,19 +111,40 @@ def reshape(self, *newshape):
+        from sympy import Derivative
+        from sympy import Dict
+        if isinstance(arg, (Iterable, Tuple, MatrixCommon, NDimArray)):
+            return type(self)([self.diff(x) for x in arg])
```
This case still cast sparse array to dense array. I have tried `derive_by_array` but `derive_by_array` itself has the same problem. Maybe an algorithm to calculate the new index after concatenating some sparse arrays, so that we can initiate a sparse array with the passed dictionary. That is to say:

[self.diff(x) for x in arg] -> [a, b, c] where a, b, c are sparse arrays. 

The question is how to calculate the index of non-zero value of a, b, c in a new dictionary, say D, so that D can represent [a, b, c].

kangzhiq commented on sympy/tensor/array/sparse_ndim_array.py:
```diff
@@ -111,19 +111,40 @@ def reshape(self, *newshape):
+        if isinstance(other, SparseNDimArray):
+            from sympy import Dict
+            dict_self = Dict(self._sparse_array)
+            dict_other = Dict(other._sparse_array)
```
The dictionaries are cast to `Dict` because SymPy 's `Dict` and Python's built-in `dict` are not fully compatible. By saying that I mean that 
```python
>>> from sympy import Dict
>>> a = Dict({3:5, 2:7})
>>> b = {3:5, 2:7}
>>> a == b
False
```
They are not equal although they have exactly the same value.

Upabjojr replied:
No casting please. There are better ways to compare.

kangzhiq commented on sympy/tensor/array/sparse_ndim_array.py:
```diff
@@ -111,19 +111,40 @@ def reshape(self, *newshape):
+        from sympy.core.compatibility import Iterable
+        from sympy.matrices.common import MatrixCommon
+        from sympy import Derivative
+        from sympy import Dict
```
Imported API may need regrouped. Waiting for some advices. :-)

kangzhiq commented on sympy/tensor/array/sparse_ndim_array.py:
```diff
@@ -109,8 +109,42 @@ def reshape(self, *newshape):
+                    s_a2[key] = deri
+            return type(self)(Dict(s_a2), self._shape)
+
+    def __eq__(self, other):
```
Since the evaluation of derivative triggers `__eq__`, this issue is handled together in this PR. I don't know if it is appropriate to do so. Waiting for some advices. :-)

Upabjojr replied:
What about temporarily ignore equality checks?

kangzhiq replied:
@Upabjojr Do you mean that we keep equality checks casting sparse array to dense array and handle the case of `__eq__` in another PR?

Upabjojr replied:
There is already a method handling __eq__ in the superclass. I think the code should remain there.

Upabjojr commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -129,34 +129,40 @@ def f(pointer):
+        from sympy import Dict
+
+        if shape is None:
+            if iterable is None:
```
is this change necessary?

kangzhiq replied:
This change is to refactor the code so that ```shape is None``` will not be repeated several times

Upabjojr replied:
OK, open a new PR next time. A separate PR on this issue would already be mergeable.

Upabjojr commented:
Tests are needed.

Upabjojr commented:
Generally, you should not create new methods for equality checks. There is already one I believe, you should probably work on that one.

Commit:
Improved derive_by_array

- Added support for derivative by array for sparse array
- Modified _eval_derivative_array in sparse array

Commit:
Removed unnecessary overrides and add tests

- Removed override method __eq__, __hash__, eval_derivative_array in
Sprase array
- Modified __eq__ in NDimArray to add case for sparse array
- Modified derive_by_array to add case for sparse array
- Added tests of __eq__ for large scale sparse array
- Added tests for derive_by_array for large scale sparse array

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
+    sparse_array = {}
+    loop_size = len(expr)
+    for i, x in enumerate(dx):
+        sparse_diff = derive_by_array(expr, x)._sparse_array
```
why is this calling derive_by_array?

kangzhiq replied:
This is because `.diff()` was slow when it comes to large scale sparse array, so I chose not to use it. Besides, `derive_by_array `, as said in the description of the function, supports the derivate by scalars, so I used it. If `.diff()` or `applyfunc` is fixed, then we can simply call `expr.diff(x)`.

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
     from sympy.matrices import MatrixBase
+    from sympy.tensor import SparseNDimArray
     array_types = (Iterable, MatrixBase, NDimArray)
-
```
why was this line remove?

kangzhiq replied:
I have seen other examples in NDimArray that there are not new lines between the `import `and the executive code, which made me feel like it is ok to remove it.

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
+                new_array = [[y.diff(x) for y in expr] for x in dx]
             return type(expr)(new_array, dx.shape + expr.shape)
         else:
-            return expr.diff(dx)
```
This line was correct. IIRC, this depends on the applyfunc method, so just fix that method.

kangzhiq replied:
I think it is a little bit compliacted to modify `applyfunc.` My initial idea is to apply the function only to the dictionary of sprase array, in this case, `diff`, it makes sense. But when it comes to some operations which change 0 to other value, e.g. `lambda x: x+1` , every value should be added by one, not only the ones in the dictionary.
So I only modified the function which triggers applyfunc.

kangzhiq replied:
I am trying with a test like this to make sure 0 is not changed:
```python
if f(0) == 0 and isinstance(self, SparseNDimArray):
        return type(self)({k: f(v) for k, v in my_dictionary.items()},self.shape)
```
But f(0) raised several errors. Do you have any idea to find a bypath for this test?

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
+
         if isinstance(dx, array_types):
-            new_array = [[y.diff(x) for y in expr] for x in dx]
+            if isinstance(expr, SparseNDimArray):
```
if both expr and dx are sparse, just use a dict comprehension (in a single line).

kangzhiq replied:
The line 213 uses dict comprehension when expr is sparse but dx is not. 
I think for the case where `dx` is an array(sparse or dense), it is necessary to calculate the new index for the derivative by each term in `dx` , which is the purpose of `_sparse_array_derive_by_array`.

Upabjojr replied:
Yes, so why not just an array comprehension?

Upabjojr replied:
`{k1+k2: v2.diff(v1) for k1, v1 in dx... dict ...items() for k2, v2 in expr... dict ...items()}`

kangzhiq replied:
OK, I got your point! Thank you!

Upabjojr commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -129,34 +129,40 @@ def f(pointer):
 
         # Construct NDimArray(iterable, shape)
         elif shape is not None:
             pass
```
if you remove "else" under this line, this code becomes useless.

Upabjojr commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -420,9 +426,19 @@ def __eq__(self, other):
+                   all(v == other._sparse_array[k] for (k, v) in self._sparse_array.items()
+                       if k in other._sparse_array)
+
+        return list(self) == list(other)
```
why not just use something similar to this? dict(...) == dict(... other ...) ?

kangzhiq replied:
I thought I am not supposed to cast one type to the other so I came up with this code. But if we can cast `Dict` to `dict`, that would indeed be much simpler. :-)
By the way, the problem of comparing `Dict` and `dict` will dispear if the test about `mutableArray.diff() == MutableArray` is changed to `mutableArray.diff() == ImmutableArray()`.

Upabjojr replied:
Just cast to dict.

Upabjojr commented on sympy/tensor/array/sparse_ndim_array.py:
```diff
@@ -109,8 +109,7 @@ def reshape(self, *newshape):
 
-        return type(self)(*(newshape + (self._array,)))
-
+        return type(self)(*(newshape + (self._sparse_array,)))
```
is this a bugfix?

kangzhiq replied:
Yes, but apprently the problem isn't resolved. I will discard this change and fix this issue in a seperated PR.

Upabjojr commented on sympy/tensor/array/tests/test_arrayop.py:
```diff
@@ -84,6 +85,13 @@ def test_derivative_by_array():
+    b = MutableSparseNDimArray.zeros(10000, 20000)
+    b[0, 0] = i
+    b[0, 1] = j
+    assert derive_by_array(b, i)._sparse_array == Dict({0: 1})
```
what about comparing it to an array?

Commit:
Temporarily resolved derive_by_array

- Temorarily fixed .diff()
- Fixed some mistakes
- Ameliorated tests

Upabjojr commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -251,10 +253,15 @@ def _eval_derivative(self, arg):
         if isinstance(arg, (Iterable, Tuple, MatrixCommon, NDimArray)):
             return derive_by_array(self, arg)
         else:
+            if isinstance(self, SparseNDimArray):
```
This method was not supposed to be modified, applyfunc should be able to handle sparse arrays efficiently.

Commit:
Ameliorated functions and removed unnecessary code

- Ameliorated derive_by_array using dict comprehension
- Removed _sparse_array_derive_by_array
- Ameliorated applyfunc for the case of sparse array
- Ameliorated __eq__ in NDimArray by casting dictionaries to dict

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -197,9 +198,20 @@ def derive_by_array(expr, dx):
+                lp = len(expr)
+                new_array = {k + i*lp: expr.diff(x)._sparse_array[k]
+                             for k in expr._sparse_array
+                             for i, x in enumerate(dx)
```
why not iterate over the sparse array? You are flattening all members and then filtering with an if on the next line.

kangzhiq replied:
This is because dx is an arraytype, not exactly a sparse array. Besides, I was allowing flatterning of `dx` because `dx` is supposed to be a sequence of non-zero values because 
```
>>> S.Zero._diff_wrt
False
```
So I think `dx` is not supposed to be a sparse array. Am I misunderstanding something here? Please let me know if so, thank you!!

kangzhiq replied:
I think I got your point! Working on it.

Upabjojr commented on sympy/tensor/array/arrayop.py:
```diff
@@ -209,7 +221,6 @@ def derive_by_array(expr, dx):
         else:
             return diff(expr, dx)
 
-
```
leave two blank lines separation in the zeroth indentation.

Upabjojr commented on sympy/tensor/array/sparse_ndim_array.py:
```diff
@@ -111,7 +111,6 @@ def reshape(self, *newshape):
 
         return type(self)(*(newshape + (self._array,)))
 
-
```
two blank lines in zeroth indentation

Upabjojr commented on sympy/tensor/array/ndim_array.py:
```diff
@@ -270,6 +274,11 @@ def applyfunc(self, f):
         """
+        from sympy.tensor.array import SparseNDimArray
+
+        if isinstance(self, SparseNDimArray) and f(sympify(0)) == 0:
```
S.Zero is the SymPy zero.

Upabjojr commented on sympy/tensor/array/tests/test_mutable_ndim_array.py:
```diff
@@ -131,6 +131,14 @@ def test_sparse():
+    a[1, 1] = 1
+    b[1, 1] = 2
+    assert a != b
+
```
test the behaviour of .applyfunc(fun) where fun(0) != 0

Commit:
Ameliorated codes and added tests

- Ameliorated codes in derive_by_array
- Added test for f(0) != 0 for applyfunc
